Improve the waiting mechanism for keeper in integration tests

azat opened the pull request:
- Add `connection_retry`/`command_retry`
- Improve `wait_until_connected` to check that keeper can establish new sessions as well

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

CI:
- https://s3.amazonaws.com/clickhouse-test-reports/72598/a7e97aa3a2dd13ba4fd5c2e37ff7299187d85bd8/integration_tests__tsan__%5B3_6%5D.html

azat committed:
tests: improve helpers for waiting for keeper to became ready

4LW commands not always enough to ensure that keeper can server regular
queries, since for regular queries it need to create session as well.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: add connection/command for KazooClient for Keeper

CI: https://s3.amazonaws.com/clickhouse-test-reports/72598/a7e97aa3a2dd13ba4fd5c2e37ff7299187d85bd8/integration_tests__tsan__%5B3_6%5D.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #74076 from azat/ClickHouse:itests/keeper-retry -->
*This is an automated comment for commit e6b7e822c1bade1b96a755096807e0e2bf85fb3a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/74076/e6b7e822c1bade1b96a755096807e0e2bf85fb3a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/73688/64174d09f1e8b24b30022e191dda387552fc9be3/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs [ClickBench](https://github.com/ClickHouse/ClickBench/) with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docker keeper image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/docker_keeper_image.html">✅ success</a></td></tr>
<tr><td>Docker server image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/docker_server_image.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/5eaad6a95417f62e5dcf6c4b06abcdf2e40e399c/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/fast_test.html">✅ success</a></td></tr>
<tr><td>Flaky tests</td><td>Checks if new added or modified tests are flaky by running them repeatedly, in parallel, with more randomization. Functional tests are run 100 times with address sanitizer, and additional randomization of thread scheduling. Integration tests are run up to 10 times. If at least once a new test has failed, or was too long, this check will be red. We don't allow flaky tests, read <a href="https://clickhouse.com/blog/decorating-a-christmas-tree-with-the-help-of-flaky-tests/">the doc</a></td><td>✅ success</td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/stateful_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stateless tests</td><td>Runs stateless functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/stateless_tests__debug__[1_2].html">✅ success</a></td></tr>
<tr><td>Stress test</td><td>Runs stateless functional tests concurrently from several clients to detect concurrency-related errors</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/stress_test__debug_.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d144f444affac79d6b43b45fc1d68c46a7766750/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

azat committed:
tests: remove unused close_keeper_socket()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: remove copy paste of get_keeper_socket()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: disable keeper sessions check for test_keeper_disks

CI: https://s3.amazonaws.com/clickhouse-test-reports/74076/21f310c99887e2d3189555aa584ba0c655087435/integration_tests__asan__old_analyzer__[6_6].html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[Integration tests (aarch64) [2/6]](https://s3.amazonaws.com/clickhouse-test-reports/74076/965732998df57b4dce744ae842640f0154b18577/integration_tests__aarch64__[2_6].html) — fail: 1, passed: 520

- #72458 

>[Integration tests (tsan) [1/6]](https://s3.amazonaws.com/clickhouse-test-reports/74076/965732998df57b4dce744ae842640f0154b18577/integration_tests__tsan__[1_6].html) — fail: 41, passed: 117

- `test_storage_kafka`

> [Integration tests flaky check (asan)](https://s3.amazonaws.com/clickhouse-test-reports/74076/965732998df57b4dce744ae842640f0154b18577/integration_tests_flaky_check__asan_.html) — Failed tests found: error: 0, passed: 652, fail: 168, skipped: 0, broken: 0, not_fail: 0

The problem is that the tests are not idempotent, and it is not that easy to do this (for some of them), so I would suggest to do this separately

azat commented:
Can someone please take a look?

antonio2368 reviewed (approved):
doesn't seem worse
